db/range_del_aggregator.cc

```diff
@@ -591,20 +589,44 @@ void RangeDelAggregator::InvalidateRangeDelMapPositions() {
   }
 }
 
-RangeDelMap& RangeDelAggregator::GetRangeDelMap(SequenceNumber seq) {
+RangeDelMap* RangeDelAggregator::GetRangeDelMapIfExists(SequenceNumber seq) {
   assert(rep_ != nullptr);
   // The stripe includes seqnum for the snapshot above and excludes seqnum for
   // the snapshot below.
+  if (rep_->stripe_map_.empty()) {
+    return nullptr;
+  }
   StripeMap::iterator iter;
   if (seq > 0) {
     // upper_bound() checks strict inequality so need to subtract one
     iter = rep_->stripe_map_.upper_bound(seq - 1);
```

Comment:
I have some fear about what will happen if the earliest snapshot for which this `ShouldDelete`'s key is visible does not yet have an entry in the map, while the snapshot just newer than it has a range tombstone covering the key. Is it possible we determine the key should be dropped due to being covered by the newer snapshot stripe's range tombstone, which would violate the current snapshot stripe's correctness?

Reply:
One solution could be pass `CompactionIterator::current_user_key_snapshot_` to `ShouldDelete` and search for that exact key.

Reply:
I'm thinking of changing the stripe map to store a `pair<RangeDelMap, size_t>` so that it has an index back into the snapshot vector. Then, all we need to do is check if the previous entry in `snapshots_` to determine whether or not the map entry we found is the one we want, or if the one we want hasn't been created.